src/Module/Search/Index.php

```diff
@@ -197,4 +196,89 @@ public static function content(array $parameters = [])
 
 		return $o;
 	}
+
+	/**
+	 * Tries to redirect to a local profile page based on the input.
+	 *
+	 * This method separates logged in and anonymous users. Logged in users can trigger contact probes to import
+	 * non-existing contacts while anonymous users can only trigger a local lookup.
+	 *
+	 * Formats matched:
+	 * - @user@domain
+	 * - user@domain
+	 * - Any fully-formed URL
+	 *
+	 * @param BaseURL $baseURL
+	 * @param string  $search
+	 * @throws HTTPException\InternalServerErrorException
+	 * @throws \ImagickException
+	 */
+	private static function tryRedirectToProfile(BaseURL $baseURL, string $search)
+	{
+		$isUrl = parse_url($search, PHP_URL_SCHEME) !== '';
+		$isAddr = preg_match('/^@?([a-z0-9.-_]+@[a-z0-9.-_:]+)$/i', trim($search), $matches);
+
+		if (!$isUrl && !$isAddr) {
+			return;
+		}
+
+		if ($isAddr) {
+			$search = $matches[1];
+		}
+
+		if (local_user()) {
+			// User-specific contact URL/address search
+			$contact_id = Contact::getIdForURL($search, local_user());
+			if (!$contact_id) {
+				// User-specific contact URL/address search and probe
+				$contact_id = Contact::getIdForURL($search);
+			}
+		} else {
+			// Cheaper local lookup for anonymous users, no probe
+			if ($isAddr) {
+				$contact = Contact::selectFirst(['id' => 'cid'], ['addr' => $search, 'uid' => 0]);
+			} else {
+				$contact = Contact::getDetailsByURL($search, 0, ['cid' => 0]);
+			}
+
+			if (DBA::isResult($contact)) {
+				$contact_id = $contact['cid'];
+			}
+		}
+
+		if (!empty($contact_id)) {
+			$baseURL->redirect('contact/' . $contact_id);
+		}
+	}
+
+	/**
+	 * Fetch/search a post by URL and redirects to its local representation if it was found.
+	 *
+	 * @param BaseURL $baseURL
+	 * @param string  $search
+	 * @throws HTTPException\InternalServerErrorException
+	 */
+	private static function tryRedirectToPost(BaseURL $baseURL, string $search)
+	{
+		if (parse_url($search, PHP_URL_SCHEME) != '') {
```

Comment:
I would prefer exiting with `return` at the top of the function when the check isn't successful. I dislike large `if` constructs that reach through the whole function.

Reply:
Fair enough.